### PR TITLE
Add FASTER NameIndex

### PR DESCRIPTION
**Pull request by timothycoleman**
Added: Stream Records and Stream Name Index

- Add Stream records
  - Index them in `$streams-created`
  - Use the index for the id -> name lookup
  - Present in $all reads and subscriptions

- Add Stream Name Index
  - Persists, catches up, truncates 
  - InMem version

Misc work:
- Move `CopyForRetry` to IPrepareLogRecord<> so that each kind of prepare can create a copy of the same type
- Some assorted renaming (sorry this ended up in the same pr)

### CODE REVIEW

**Comment by timothycoleman** [comment]
closing this for the second to avoid lots of upcoming ci activity

**Review by pvanbuijtene — APPROVED** [review]

File: src/EventStore.Core.Tests/Services/Storage/DeletingStream/when_writing_few_prepares_on_same_event_number_and_commiting_delete_on_this_version_read_index_should.cs [hunk 1]
```diff
@@ -17,9 +17,9 @@ public class
 		protected override void WriteTestScenario() {
 			long pos;
 			string stream = "ES";
-			_streamNameIndex.GetOrAddId(stream, out var streamId, out _, out _);
+			GetOrReserve(stream, out var streamId, out pos);
 
-			var prepare1 = LogRecord.SingleWrite(_recordFactory, WriterCheckpoint.ReadNonFlushed(), // prepare1
+			var prepare1 = LogRecord.SingleWrite(_recordFactory, pos, // prepare1
```
Comment (shaan1337) [inline src/EventStore.Core.Tests/Services/Storage/DeletingStream/when_writing_few_prepares_on_same_event_number_and_commiting_delete_on_this_version_read_index_should.cs:22]: @timothycoleman doesn't the position change between the prepare1, prepare2, etc. writes?

Comment (timothycoleman) [reply src/EventStore.Core.Tests/Services/Storage/DeletingStream/when_writing_few_prepares_on_same_event_number_and_commiting_delete_on_this_version_read_index_should.cs:22]: yep, the call to Writer.Write updates it between each

**Review by shaan1337 — APPROVED** [review]
great work! :+1:

File: src/EventStore.Core.Tests/Services/Storage/DeletingStream/when_writing_few_prepares_on_same_event_number_and_commiting_delete_on_this_version_read_index_should.cs (hunk 1 above)
Comment (shaan1337) [reply src/EventStore.Core.Tests/Services/Storage/DeletingStream/when_writing_few_prepares_on_same_event_number_and_commiting_delete_on_this_version_read_index_should.cs:22]: ohh missed that, thanks! :)

**Pull request merged**